### PR TITLE
Quick modernisation of the "get started with charm testing" how-to guide

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
An update for: https://juju.is/docs/sdk/get-started-with-charm-testing

I have been linking to this page from all the how-to guides without actually reading it, and so did not notice how very outdated it is :disappointed:.

I think we'll want to do a larger reworking of this - particularly as we transition away from Harness to Scenario, and we should look at how it overlaps with the tutorial on testing.

However, in the meantime, I think it would be good to at least remove the links to things that are outdated so that it's immediately usable.

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@tmihoc could you please review this when you have time? Thanks!

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 1]
```diff
@@ -1,10 +1,9 @@
 Testing charm code is an essential part of charming. Here we will see how to get started with it. We will look at the templates we have available and the frameworks we can use to write good unit, integration, and functional tests.
 
-
 Prerequisites:
 - knowledge of testing in general
-- knowledge of juju and charms
-- knowledge of the juju models and events, esp. the data involved in a charm's lifecycle (e.g. see [Talking to a workload control flow from A to Z](/t/6161)
+- knowledge of Juju and charms
+- knowledge of the Juju models and events, esp. the data involved in a charm's lifecycle (e.g. see [Talking to a workload control flow from A to Z](/t/6161)
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:7]: After the link, closing bracket for the example missing.

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 2]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
-    export PYTHONPATH="lib:src"                       
-else                                                  
-    export PYTHONPATH="lib:src:$PYTHONPATH"           
-fi                                                    
-                                                      
-flake8                                                
-coverage run --branch --source=src -m unittest -v "$@"
-coverage report -m                                    
+```shell
+charmcraft init --profile=machine
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:24]: This feels out of scope for the doc.

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 3]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
+│   └── unit
+│       └── test_charm.py
+└── tox.ini
+```
 
-[note]`template-operator` is in practice a k8s charm template. Its example code is that of a kubernetes charm, and its CI will run the integration tests on `microk8s`. If you're developing a machine charm, you are going to have to change all that.[/note]
+There are also profiles for `kubernetes` and for building charms for apps developed with popular frameworks such as Django and Flask.
 
-In the rest of this tutorial we will assume that our project is set up with `template-operator`.
+> See more: [ Write your first Kubernetes charm for a Flask app](https://juju.is/docs/sdk/write-your-first-kubernetes-charm-for-a-flask-app)
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:51]: We have all of this info on charmcraft init and profiles here: https://juju.is/docs/sdk/set-up-a-charm-project 

If we're missing details, we should add them there rather than trying to redefine here.

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 4]
```diff
@@ -17,97 +16,85 @@ What you will learn:
   - Where can you apply functional testing?
 - How to automate this in a CI pipeline.
 
+# Charmcraft Profiles
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:20]: Here and everywhere else: Title should follow sentence case. (It's what we do everywhere else at Canonical.)

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 5]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
-In the rest of this tutorial we will assume that our project is set up with `template-operator`.
+> See more: [ Write your first Kubernetes charm for a Flask app](https://juju.is/docs/sdk/write-your-first-kubernetes-charm-for-a-flask-app)
 
 # Unit Testing
 
-
 ## A charm as an input -> output function
-In production, a charm is a stateless (modulo stored state) object that comes to life when the juju agent decides to execute it with a given context (we call that an `event`). 
+
+In production, a charm is an object that comes to life when the Juju agent decides to execute it with a given context (we call that an `event`).
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:56]: This is true in general, not just in production. And a HTG on testing feels way too late to introduce this info. To the extent that we want to give this info again here as an intro, it should be worded as a reminder, not as text introducing this info for the first time.

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 6]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
 
-The charm code executes and typically produces side-effects aimed at its workload (i.e. it writes files to a disk, runs commands on a system, or reconfigures a process) or at other charms it integrates with (i.e. it writes relation data). We call this 'operating' a workload, and that is what a charm is meant to do.
-The ways in which a charm operates can be roughly categorized as:
+The charm code executes and typically produces side-effects aimed at its workload (for example: it writes files to a disk, runs commands on a system, or reconfigures a process) or at other charms it integrates with (for example: it writes relation data). We call this 'operating' a workload, and that is what a charm is meant to do. The ways in which a charm operates can be roughly categorised as:
  
 - system operations (e.g. kill a process, restart a service, write a file, emit to a syslog server, make a HTTP request)
-- cloud operations (e.g. deploy a kubernetes service, launch a VM)
+- cloud operations (e.g. deploy a Kubernetes service, launch a VM)
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:69]: I'm confused by "cloud operations". In the Juju paradigm the charm never talks to the cloud directly, right?

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 7]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
+> See more [`ops.testing.Harness`](https://ops.readthedocs.io/en/latest/#ops.testing.Harness)
 
-The harness is the 'mocker' for most inputs and outputs you will need. Where a live charm would gather its input through context vars and calls to the juju api (by running the hook tools), a charm under unit-test will gather data via a mocked backend managed by the harness. Where a live charm would produce output by writing files to a filesystem, the harness exposes a mock filesystem the charm will be able to interact with without knowing the difference. More specific outputs, however, will be need to be mocked individually
+Harness is the 'mocker' for most inputs and outputs you will need. Where a live charm would gather its input through context variables and calls to the Juju api (by running the hook tools), a charm under unit test will gather data via a mocked backend managed by Harness. Where a live charm would produce output by writing files to a filesystem, Harness exposes a mock filesystem the charm will be able to interact with without knowing the difference. More specific outputs, however, will be need to be mocked individually.
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:96]: will need to be*

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 8]
```diff
@@ -17,97 +16,85 @@ What you will learn:
...
 
-A typical test with the harness will look like this:
+A typical test with Harness will look like this:
  
 - set up:
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md]: set things* up (sounds more natural)

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 9]
```diff
@@ -244,89 +190,79 @@ Things you typically want to test with integration tests:
...
     assert charm_operates_correctly()  
 ```
 
-`ops_test` has, of course, API for all inverse operations: remove an app, scale it down, remove a relation...
+`python-libjuju` has, of course, an API for all inverse operations: remove an app, scale it down, remove a relation...
 
 A good integration testing suite will check that the charm continues to operate as expected whenever possible, by combining these simple elements.
 
-
 # Functional Testing
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md]: Shall we call this "Scenario testing", to align with what we call it everywhere else?

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 10]
```diff
@@ -244,89 +190,79 @@ Things you typically want to test with integration tests:
...
-Some charms represent their workload by means of an object-oriented wrapper, which mediates between operator code and the implementation of operation logic.
-In such cases, it can be useful to add a third category of tests, namely functional tests, that black-box test that workload wrapper without worrying about the substrate it runs on (the charm, the cloud, the machine or pod...).
-For an example charm adopting this strategy, see [parca-operator](https://github.com/jnsgruk/parca-operator).
 
+Some charms represent their workload by means of an object-oriented wrapper, which mediates between operator code and the implementation of operation logic. In such cases, it can be useful to add a third category of tests, namely functional tests, that black-box test that workload wrapper without worrying about the substrate it runs on (the charm, the cloud, the machine or pod...).
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:249]: scenario* tests

File: sdk/how-to-guides/getting-started-with-testing.md [hunk 11]
```diff
@@ -244,89 +190,79 @@ Things you typically want to test with integration tests:
...
+For an example charm adopting this strategy, see [parca-operator](https://github.com/jnsgruk/parca-operator). Nowadays, the preferred tool to do functional testing is Scenario.
```
Comment (tmihoc) [inline sdk/how-to-guides/getting-started-with-testing.md:251]: There is no other way, right? So this whole section should just become about scenario.

**Review by tmihoc — COMMENTED** [review]
Thanks for updating the links! I left some comments but I think you should feel free to ignore anything that's more than a typo or a comma, because I think we need to rethink this doc more generally, because:

This doc has long been on my mind. As everything written by Pietro, it's full of gems. As everything written by someone who knows what they're talking about, it tries to be self-sufficient and, as a result, fails to be modular in a way that would promote good integration with the other docs and good maintainability. (For example, as you noticed, the "Get started with" has a tutorial vibe, and the bits that talk about charmcraft init are already covered in a separate how-to guide. Plus we probably should have both this doc and the testing reference doc.) Overall, I think we should rethink how we deliver this content in general.

**Comment by tonyandrewmeyer** [comment]
> Thanks for updating the links! I left some comments but I think you should feel free to ignore anything that's more than a typo or a comma, because I think we need to rethink this doc more generally, because:
> 
> This doc has long been on my mind. [...] Overall, I think we should rethink how we deliver this content in general.

I completely agree that this needs to be reworked and we should rethink how this content is delivered. My concern is that at the moment does things like point to archived/outdated repos, has very outdated content on Harness, and so on. This was a super-quick update to get it up-to-date before the reworking/rethinking that will probably take more time - particularly since all the how-to guides are linking to this page.

File: sdk/how-to-guides/getting-started-with-testing.md (hunk 2 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/getting-started-with-testing.md:24]: I was trying to do a minimal change (before the rethinking/refactoring project), and so just swapped out the previous (outdated) template content for charmcraft content.

File: sdk/how-to-guides/getting-started-with-testing.md (hunk 3 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/getting-started-with-testing.md:51]: Same as above, it currently has a bunch of content from the outdated template repo, so this was an attempt at a 1-to-1 swap out.

(I do agree that when this is refactored/rethought it shouldn't have this duplication).

File: sdk/how-to-guides/getting-started-with-testing.md (hunk 5 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/getting-started-with-testing.md:56]: I was just removing "stateless (modulo stored state)" as that felt confusing. I do agree that this needs reworking, but thought that would be in the bigger rethink/refactor.

File: sdk/how-to-guides/getting-started-with-testing.md (hunk 6 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/getting-started-with-testing.md:69]: Yes, this does seem odd (but all I did was capitalise "Kubernetes" :smile:). Charms manage services - I'm not sure I would say "deploy", and indeed it's really Juju that is launching VMs.

File: sdk/how-to-guides/getting-started-with-testing.md (hunk 9 above)
Comment (tonyandrewmeyer) [reply sdk/how-to-guides/getting-started-with-testing.md]: This section is a bit odd, because it's linking out to Jon's parca-operator repo as an example of functional testing, but then says to do Scenario tests, and those aren't the same. Jon's functional tests are testing the Parca lib, don't use Harness or Scenario, and I don't think you would rewrite them to do so (no Juju events are involved).

"unit", "functional", "integration" and other test categories are somewhat loosely defined in the industry, which makes this kind of thing tricky.

I think this should be left for the refactor/rethinking, and content like the [Best Practices](https://discourse.charmhub.io/t/best-practices-for-testing-charms/15172) guide could maybe replace this? For testing charms, our current recommendation is to do unit tests (today with both Harness and Scenario, but any-moment-now-(tm) just Scenario) and integration tests (today, with pytest-operator and pylibjuju). If you're testing a charm lib, that probably deserves its own guide, and it depends a lot on whether it's isolated from Juju (like Jon's one) or emits custom events.

**Comment by tonyandrewmeyer** [comment]
@tmihoc I'll update it as it is now - it's at least reasonably up-to-date then.

I think everything remaining is better left to the larger rethink/refactor/rework. For what it's worth, the Charm-Tech team are planning on having a team discussion in The Hague about the path forward with testing (particularly beyond unit testing, since we're settled on Scenario for that now), so the rework might be a good follow-up from that. Maybe we could combine that with the "move the tutorial tests to the individual chapters" task we tentatively have for the 25.04 roadmap to make it a larger "improve the docs on charm testing" item.

(cc @benhoyt).
